session/send.go

```diff
@@ -68,16 +75,57 @@ type Message struct {
 	IsDecoy bool
 }
 
+func (s *Session) WaitForSent(msgId *[cConstants.MessageIDLength]byte) error {
+	s.log.Debug("Waiting for message to be sent.")
+	s.mapLock.Lock()
+	msg, ok := s.messageIDMap[*msgId]
+	if !ok {
+		return fmt.Errorf("[%v] Failure waiting for reply, invalid message ID.", msgId)
+	}
+	if msg.Sent {
+		return nil
+	}
+	waitCh, ok := s.waitSentChans[*msgId]
+	if !ok {
+		return fmt.Errorf("[%v] Failure waiting for reply, invalid message ID.", msgId)
+	}
+	s.mapLock.Unlock()
+	<-waitCh.Out()
+	s.log.Debug("Finished waiting. Message was sent.")
+	return nil
+}
+
 // WaitForReply blocks until a reply is received.
-func (s *Session) WaitForReply(msgId *[cConstants.MessageIDLength]byte) []byte {
+func (s *Session) WaitForReply(msgId *[cConstants.MessageIDLength]byte) ([]byte, error) {
 	s.log.Debugf("WaitForReply message ID: %x\n", *msgId)
+
+	err := s.WaitForSent(msgId)
+	if err != nil {
+		return nil, err
+	}
+
 	s.mapLock.Lock()
-	replyLock := s.replyNotifyMap[*msgId]
+	waitCh, ok := s.waitChans[*msgId]
+	if !ok {
+		return nil, fmt.Errorf("[%v] Failure waiting for reply, invalid message ID.", msgId)
+	}
+	msg, ok := s.messageIDMap[*msgId]
+	if !ok {
+		return nil, fmt.Errorf("[%v] Failure waiting for reply, invalid message ID.", msgId)
+	}
+	s.log.Debug("reply eta is %v", msg.ReplyETA)
 	s.mapLock.Unlock()
-	replyLock.Lock()
-	s.mapLock.Lock()
-	defer s.mapLock.Unlock()
-	return s.messageIDMap[*msgId].Reply
+
+	select {
+	case event := <-waitCh.Out():
+		e, ok := event.(MessageReplyEvent)
+		if ok {
+			return e.Payload, nil
+		}
+	case <-time.After(msg.ReplyETA + (msg.ReplyETA / 2)): // XXX
```

Comment:
Yeah, I tried something like this and my thought is that the timeout needs to be adjusted based on the fetch interval. See above comments.